### ISSUE TITLE
/public/ folder discussion

### ISSUE THREAD

**Issue by objecttothis** [opensourcepos/opensourcepos#1491]
### New Feature / Enhancement
Hi everyone.  Being new to the current version of ospos, there are all kinds of new things to get familiar with.  One of the first things is needing to go to the public folder to access the app.  From seeing the sheer number of repeat questions on it, it appears I'm not the only one that is thrown by that one.  From searching the closed issues, this seems to be a feature introduced with composer.  I have a suggestion that would fix this problem and eliminate all the "I get HTTP 403 when I go to http://url_to_ospos/" questions with one line of code.

An index.html with `<meta http-equiv="refresh" content="0; url=./public/" />` in the root would redirect to public and everybody is happy!  Thoughts?

### Installation information

- OSPOS version is: 3.1.0
- OSPOS git commit hash is: 
- PHP version is:   (e.g. 5.5, 5.6, 7.0, 7.1) 5.6.31
- MySQL or MariaDB version is:   (e.g. MySQL 5.5, MySQL 5.6, MySQL 5.7, MariaDB 10.0, MariaDB 10.1, MariaDB 10.2) 5.7
- OS and version is:   (e.g. CentOS 6.9, Ubuntu 16.4, Windows 10) FreeBSD 11.0
- WebServer is:   (e.g. Apache 2.2, Apache 2.4, Nginx 1.12, Nginx 1.13) Apache 2.4
- (If applicable) Installation package for the LAMP/LEMP stack is:   (e.g. WAMP, XAMPP) FAMP

**Comment by jekkos** [comment]
Might be better to handle this on webserver level.. like adding a rule in apache .htaccess

**Comment by objecttothis** [comment]
@jekkos Yes, I looked into using an apache rule but there's a problem.  We would need to add the standard index.html into the root directory then add the redirect rule below to the .htaccess in the root directory.

`Redirect 301 /index.html /public/`

the above code works, but if the user is using a subdirectory, they will need to change the line to 

`Redirect 301 /{subdirectoryname}/index.html  /{subdirectoryname}/public/`

Users needing to change their .htaccess files already creates more issues being created.  That too would be resolved by an installer that could detect and customize based on install location.

A solution that should be more secure than the HTML redirect but not contain any of the potential to need to modify .htaccess files is to do the redirect with php. An index.php file in the root directory with the following two lines of code would do the trick

`header("Location: ./public/",TRUE,301);`
`die();`

Thoughts?

**Comment by jekkos** [comment]
Can't we just use a catch all pattern instead of index.html, think that way we can fit both cases in one line? That's most of the time how these rewrite rules are done. Ofcourse you'll need to enable mod_rewrite (which might or might not be done by default depending on the distribution).

**Comment by daN4cat** [comment]
At the time I spent some time trying to get an .htaccess right that could forward to the public/ but I always ended up with infinitive redirection loop issues, so I gave up.

**Comment by jekkos** [comment]
Think the point is that you only need to match the first (parent) dir? I tried something here not sure if it will work..

**Comment by jekkos** [comment]
Think we should try to match REQUEST_URI in parent directory and see if it has the public url part in it. 
Better even, we might suggest to symlink our public directory into an apache DocumentRoot. I also did this in docker as follows

`ln -s ospos/public /var/www/html`while html being the apache 'public' folder. This way the files that have to be served are present only in the webroot while the rest remains where it is. But this solution might not be a good one for windows users (no symlinking there)

**Comment by objecttothis** [comment]
OK, in the Apache 2.4 section I added the following to the .htaccess file in the root directory

`RewriteEngine On
RewriteCond %{REQUEST_URI} !^/public$
RewriteRule ^(.*)/$ /public/ [R=301,L]`

This seems to be working on my build, but I'm not in documentroot but a subdirectory.  Can someone test this on their build and something in a documentroot directory to make sure it's redirecting properly?
As @jekkos this means we have a dependency on mod_rewrite being loaded in order for it to work.  I think this is why I prefer the php option or just using the Redirect 301 line I suggested above.  The index.html I was suggesting is the one that exists in every other subdirectory, so I don't think that being their is a security risk.

**Comment by daN4cat** [comment]
Yep the !^/public$ was probably the rule I missed at the time to avoid the redirection loop.

Also we should cover Apache 2.2 case. I bet that rule applies to both.

**Comment by daN4cat** [comment]
I submitted the change to master.

**Comment by objecttothis** [comment]
So, this redirect had not been working for me.  I finally figured out why.  For some reason the code originally seemed to work, but no longer did on nginx and apache 2.4 builds.  I am submitting a pull request for the corrected code.  If we could have a few people test the corrections.  I also introduced a check to confirm mod rewrite is enabled.

# redirect to public page
<IfModule mod_rewrite.c>
RewriteEngine On
RewriteCond %{REQUEST_URI} !^public$
RewriteRule ^(.*)$ %{REQUEST_URI}public/ [R=301,L]
</IfModule>

**Comment by jekkos** [comment]
The rule does not seem perfect and still yields an infinite redirect if it contains text before the public part here

**Comment by daN4cat** [comment]
You mean if it's in a subfolder?

**Comment by objecttothis** [comment]
@jekkos, can you give a few examples so that I can recreate and fix?

**Comment by jekkos** [comment]
if I for example enter the following

url/sometingpublic 
I get an infinite redirect. I think in that case it just keeps matching the !^public$ rewriteCond and so keeps appending the public (seeing a too many redirects in that case).

But it does work for the case in which you just go to the domain itself (then redirects to the /public beneath)

**Comment by objecttothis** [comment]
So in the example you gave the public folder is actually located at url/public or somewhere else? If I understand correctly you are saying that with the install at www.nothing.com/public/ and typing in www.nothing.com/public starts the infinite redirect but typing in www.nothing.com works ok.

**Comment by mckaygerhard** [comment]
hi @objecttothis this its not the firts time that a issue are related to infinite redirect.. i has same behaviour and solver using a index.php file as i posted previously in other issues, it's a error assumed that if worked for you worked for others..

**Comment by objecttothis** [comment]
Also I do not alter the htaccess file with directory locations which was the former protocol.

**Comment by objecttothis** [comment]
@mckaygerhard, you didn't understand me. I didn't say this is not an issue. I said it's working for me so I need to see what is different between installations so that I can recreate and fix the issue. If you'd like to post a link to your fix that includes details of what you did to fix it I'll take a look.

**Comment by daN4cat** [comment]
Chap, we reason in finding the root cause before hacking stuff in. Let @objecttothis do his work as he's on top of that.

**Comment by objecttothis** [comment]
Am I missing a conversation here?  I'm trying to download the latest bintray build to test this on a clean server and it says that the most recent one is 4-months old.  When I try downloading it, I get a Forbidden error message.

**Comment by objecttothis** [comment]
Nevermind, it's been reported in #2133

**Comment by objecttothis** [comment]
OK, I was able to replicate the issue on a LAMP stack running Apache 2.4 with the installation in the root directory of public_html.  It works correctly when installed into a subdirectory such as /public_html/pos/... but not in a root directory.

I'm glad to work on this, but as much as my patience is wearing thin for @mckaygerhard 's approach to comments, the point made is worth discussing.  The mod_rewrite rules we are using, even when working correctly, are specific to Apache.  Those running other than that will run into support issues.  So, if we want to limit installations to using *AMP server installations, we can, but if we want to increase compatibility then we may want to solve the redirect issue at the PHP level rather than .htaccess.  Regardless, this is a design decision and we need to get it working in its current form anyway.  Let me work on it and I will post the pull request for testing.

**Comment by mckaygerhard** [comment]
@objecttothis i understand you, but stay and put in shoes of a normal user that dont have enought knowledge to explain a VERY COMMON problem (very very reported here)

please take yourselft all of you and see your responses to users.. do you ?

**Comment by objecttothis** [comment]
@mckaygerhard my issue isn't usually with **what** you say, it's with **how** you are saying it.  Either change how you are saying things or you will get banned.

**Comment by WebShells** [comment]
@mckaygerhard no offense at all, but I agree with @objecttothis its the way you say things...

**Comment by mckaygerhard** [comment]
ok the threath again turn off topic, ok but that kind of response happened when a user (not only me) said sometihng that not like ... inlusivelly if are true, o not post anymore off topic, **i'm taking about here about that the redirection problem was VERY VERY reported** and was **very very silenced**

**Comment by WebShells** [comment]
Consider banning guys 🤦‍♂️🤣

**Comment by mckaygerhard** [comment]
it' a despotic response @WebShells

**Comment by WebShells** [comment]
Its a sarcastic reponse @mckaygerhard

**Comment by SteveIreland** [comment]
I hope you don't mind, but I've locked the thread to collaborators only so that it can return to being on-topic.

**Comment by objecttothis** [comment]
Not at all.

**Comment by daN4cat** [comment]
There is a documented solution for nginx not relying on Apache. Of course there are other web servers but we can't cover all possible scenarios.
If this index.php redirect does the trick it's welcome but I need to see it before saying yes or no. We need to be careful about security, performance impact of redirection happening at php level and etc.

**Comment by objecttothis** [comment]
Agreed that security is important.  As for the various web servers, I think that we should make reasonable attempts at supporting the major ones (Apache, nginx, lighttpd, IIS) without sacrificing performance and security.  This is just my opinion and as long as Apache is supported I'm fine, but if we have something bulletproof that covers the others it will increase the use of the software which, theoretically, will draw more developers to help with the project.  I'll get working on fixing the redirect as it currently stands then work on the requested information for the php level redirect.

**Comment by objecttothis** [comment]
Deleted comments to reduce clutter to this thread.  No offense intended.

**Comment by daN4cat** [comment]
I'm fine with finding a solution, please go ahead.

**Comment by jekkos** [comment]
@objecttothis ok glad to hear you can reproduce it. In fact I tried for a little while yesterday to get it fixed because as you said, it does work for standard cases, but probably will go behave weirdly in other not so standard configured setups.

Personally I like nginx more at the moment due to the simplicity by which in can be configured. I agree that finding a general solution that does not require configuring a webserver would be better, but not sure if that would be feasible.
I did a sneak peek at CI 4 and there they also move the index.php inside a public/ folder. Maybe we can check how they handle the case then and stick with that?

**Comment by daN4cat** [comment]
I noticed the same with CI4, I briefly looked around to see how they addressed the installation issue because of the same public/index.php but couldn't find any.

**Comment by objecttothis** [comment]
OK, let me make sure I'm understanding the problem well, since I seemed to have fixed part of it without understanding the difference.  Please tell me what I'm missing:

-There are a few different scenarios that we need to account for: 

1. Install is in the root directory (www.example.com) and we want .htaccess to redirect automatically to the public directory (www.example.com/public/).  

2. Install is in a subdirectory (www.example.com/pos) and we want .htaccess to redirect automatically to the public directory (www.example.com/pos/public/).

3. Regardless of the install location, when other directories are navigated to (www.example.com/vendor/), we want an automatic redirect back to the public directory (www.example.com/public/).

It's that third scenario that is failing.  Earlier when I said that I was able to recreate it, what I did was take a fresh download of the 3.2.3 release and plop it in a root directory and later in a subdirectory with only a database and no configuration done and what I found was the third scenario above was happening because a Server 500 error was happening and it couldn't redirect properly to /500.shtml so it was in an infinite loop.  Further, because my file permissions were incorrect on a few folders (that was generating the 500 error) I couldn't even get it to stop the infinite loop.  As soon as I configured the config and database file, scenario 1 and 2 above began working with no edits to the .htaccess file in /public/

So, with everything configured the current .htaccess file causes problems with scenario 3.  www.example.com/vendor becomes http://www.example.com/vendorpublic/public/public/public/.../public/

For clarity, I need to know if that's the only issue or if there are other issues that I am missing.

**Comment by objecttothis** [comment]
So our code in the root .htaccess file is currently
`<IfModule mod_rewrite.c>`
    RewriteEngine On`
    RewriteCond %{REQUEST_URI} !^public$`
    RewriteRule ^(.*)$ %{REQUEST_URI}public/ [R=301,L]`
  </IfModule>`
- Which says if the requested URI does not contain "public" then rewrite the URI to add public.
- What we need it to say is "if the requested URI does not contain "public" then rewrite the URI to be the root install directory and add public.

**Comment by daN4cat** [comment]
There is the case of subdomain too pos.example.com. Just to complete the scenarios.

**Comment by objecttothis** [comment]
@daN4cat Thank you for that.  I will add that to the scenario options to test our solution with.

**Comment by objecttothis** [comment]
Design Decision: Are we wanting to keep rewriting external redirects (www.example.com --> www.example.com/public/) or are we wanting to have a silent redirect? (www.example.com shows the content of www.example.com/public/ but the URL stays the same)? In both cases a redirect is happening.  I believe the later is better in terms of obfuscation.

**Comment by objecttothis** [comment]
OK, I think this will do it, but let me test with subdomains before submitting the pull request for people to test with.

`  <IfModule mod_rewrite.c>`
`    RewriteEngine On`
`    RewriteCond %{REQUEST_URI} !^public$`
`    RewriteRule "^(.*)$" "/public/" [R=301,L]`
`  </IfModule>`

**Comment by objecttothis** [comment]
OK, pull request created #2144 .  I tested the change on a clean install that had been configured but with no data. 

As per the pull request when testing PLEASE clear your browser cache/history and restart your browser. Most browsers have a bad habit of caching errors and then you don't know if you are getting a cached error or the current results.

**Comment by objecttothis** [comment]
I think scenario 2 may not be working correctly.  can someone confirm?

**Comment by objecttothis** [comment]
@dan4cat, thank you for that tip.  I'll go back and retest with incognito mode.  I think that I have more work to do.

**Comment by objecttothis** [comment]
the Firefox Developer Edition InPrivate Browsing mode is not much better than without it. Bummer.  I guess I'll need to use something else for testing.

**Comment by jekkos** [comment]
Ok I tried to fix it myself but probably ran into this browser caching behavior (after which I thought I couldn't get it fixed and posted this issue). Will try to test tonight.

**Comment by objecttothis** [comment]
I found this tool for debugging htaccess.  It seems to resolve caching issues, but we will see how well it does with providing results that work across the board.

https://htaccess.madewithlove.be/

To further help with our testing, let me rephrase our criteria.  There are three install types:
- Installed to root directory (www.example.com)
- Installed to subdomain (pos.example.com)
- Installed to subdirectory (www.example.com/pos)

In each install type the following tests need to pass:
- navigate to a file name (www.example.com/dir/foo.html) should redirect to the public folder.
- navigate to a non-accessible folder (www.example.com/vendor/) should redirect to the public folder.
- Requested URI contains or does not contain trailing slash / should have no impact on the redirection.
- navigate to a non-existing location (www.example.com/not/a/folder/) should redirect to the public folder.

If there are any other use cases that I'm missing please let me know.

**Comment by objecttothis** [comment]
So part of the limitations that we are running into with this is how to differentiate from www.example.com/pos and www.example.com/vendor.  In the first example, the behavior we expect is that the RewriteRule would redirect to www.example.com/pos/public but in the second we would expect the RewriteRule to redirect to www.example.com/public.  If someone can come up with a way to test for this using mod_rewrite, I'm all ears, but without writing a rewriterule for each possible variation I think we may be better off redirecting using PHP.

**Comment by daN4cat** [comment]
There could be cases where the subdomain is resolved to domain/path. Usually it's an extra rule in .htaccess, so not sure that changes something here.

**Comment by objecttothis** [comment]
@daN4cat do you mean for example something like http://pos.example.com/subdir/ being the install folder and expecting that to redirect to http://pos.example.com/subdir/public/ ?  So far in my testing, I haven't had a difficult time with that scenario.

**Comment by objecttothis** [comment]
So, the source that I have working on the htaccess tester that I posted earlier is as follows:

`  <IfModule mod_rewrite.c>`
`    RewriteEngine On`
`    RewriteCond %{REQUEST_URI} !^public$`
`    RewriteRule ^(.*)/$ //%{HTTP_HOST}/$1/public/ [R=301,L]`

`    RewriteCond %{REQUEST_URI} !^public$`
`    RewriteRule ^(.*)$ //%{HTTP_HOST}/$1/public/ [R=301,L]`
`  </IfModule>`

However, that code only accounts for the basic installations (subdomain, root and subdirectory).  It does not account for a file name in the REQUEST_URI, an inaccessable directory or a not existent directory.  Further, when I take that code and paste it into the two different server implementations that I have access to (Apache 2.4 and easy::Apache) I get different behavior.  It is a big crapshoot at this point.

**Comment by daN4cat** [comment]
I meant http://pos.example.com/ redirected to http://www.example.com/pos

It could be all fine but just wanted to add the scenario.

**Comment by daN4cat** [comment]
Ah cool, having inconsistent behaviours is great fun... 
At least they are consistent in being inconsistent.

**Comment by objecttothis** [comment]
I'm going to be done for the night.  I do have access to the .conf file on one of the two servers, so I'll probably go with adding a LogLevel directive to get a feel for how the server is treating the code.

**Comment by daN4cat** [comment]
Merged
